prestoadmin/server.py

```diff
@@ -82,11 +85,218 @@ def new_sysnode_processor(node_info_rows):
+                pass
+        return None
+    finally:
+        sys.stderr = sys.__stderr__
+
+    return download_file_path
+
+
+def _download_rpm(url_response, download_file_path):
+    content_length = _get_content_length(url_response)
+    print 'Downloading rpm from %s\n' \
+          'This can take a few minutes.' % url_response.geturl()
+
+    with open(download_file_path, 'wb') as local_file:
+        bytes_read = 0
+        block_size = 16 * 1024 * 1024
+        while True:
+            download_buffer = url_response.read(block_size)
+            if not download_buffer:
+                break
+            bytes_read += len(download_buffer)
+            local_file.write(download_buffer)
+            if content_length:
+                _print_download_status(bytes_read, content_length)
+        print "Downloaded %d bytes." % bytes_read
+
+    print 'Rpm downloaded to: %s' % download_file_path
+    return download_file_path
+
+
+def _find_or_download_rpm(url, version=None):
+    """
+    Args:
+        url:      The url of the presto rpm to be downloaded.
+        version:  An optional version number.
+                  If the server doesn't respond with the file name that is being
+                  requested, this allows the downloaded file to have the correct
+                  version attached to its name (presto-server-rpm-'version'.rpm)
+                  rather than the default name
+
+    If downloading the presto rpm at the given url would overwrite an existing rpm,
+    this function returns the path to the existing rpm. However, if the rpm that
+    would be downloaded takes the default rpm name, it will overwrite the existing
+    rpm because there is no way to know if the default rpm name is of the same version
+    as the requested rpm.
+
+    Returns:
+        Upon success, the path to the downloaded or found presto rpm
+        Upon failure, None
+    """
+    try:
+        url_response = urllib2.urlopen(url)
+    except ValueError:
+        return None
+
+    if not _check_good_response_status(url_response.getcode()):
+        return None
+
+    download_file_name = _get_download_file_name(url_response, version)
+    download_file_path = os.path.join(DOWNLOAD_DIRECTORY, download_file_name)
+
+    downloaded_rpm_path = _find_downloaded_rpm(download_file_name)
+    if downloaded_rpm_path and download_file_name != DEFAULT_RPM_NAME:
+        return downloaded_rpm_path
+
+    return _download_rpm(url_response, download_file_path)
+
+
+def _find_or_download_rpm_version_facebook(rpm_version):
+    download_url = 'http://search.maven.org/remotecontent?filepath=com/facebook/presto/' \
+                   'presto-server-rpm/' + rpm_version + '/presto-server-rpm-' + \
+                   rpm_version + '.rpm'
+    return _find_or_download_rpm(download_url, rpm_version)
+
+
+def _find_or_download_rpm_version_teradata(rpm_version):
+    abort('Finding and downloading teradata presto rpms by version is currently not supported.\n'
```

Comment:
Why do we expose this at all? Is there a plan in place to make this possible without going through the teradata website and filling in that form?